service/history/workflow/mutable_state_impl.go

```diff
@@ -5263,6 +5301,64 @@ func (ms *MutableStateImpl) closeTransactionUpdateTransitionHistory(
 	return nil
 }
 
+func (ms *MutableStateImpl) closeTransactionTrackLastUpdatedVersionedTransition(
+	transactionPolicy TransactionPolicy,
+) error {
+	if transactionPolicy != TransactionPolicyActive {
+		// TODO: replication/standby logic will need a different way for updating LastUpdatedVersionedTransition
+		// when reapplying history, especially when history replication tasks got batched.
+		return nil
+	}
+
+	if len(ms.executionInfo.TransitionHistory) == 0 {
+		// transition history is not enabled
+		return nil
+	}
+
+	transitionHistory := ms.executionInfo.TransitionHistory
+	currentVersionedTransition := transitionHistory[len(transitionHistory)-1]
+	for _, activityInfo := range ms.updateActivityInfos {
+		activityInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, timerInfo := range ms.updateTimerInfos {
+		timerInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, childInfo := range ms.updateChildExecutionInfos {
+		childInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, cancelInfo := range ms.updateRequestCancelInfos {
+		cancelInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+	for _, signalInfo := range ms.updateSignalInfos {
+		signalInfo.LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	// signalRequestedIDs is a set in DB, we don't have a place to store the lastUpdateVersionedTransition for each
+	// signal requestedID.
+	// Deletion of signalRequestID is not replicated today, so we can even drop the check on deleteSignalRequestedIDs
+	if len(ms.updateSignalRequestedIDs) != 0 || len(ms.deleteSignalRequestedIDs) != 0 {
+		ms.executionInfo.SignalRequestIdsLastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	for updateID := range ms.updatedUpdateInfosVersionedTransition {
+		ms.executionInfo.UpdateInfos[updateID].LastUpdatedVersionedTransition = currentVersionedTransition
+	}
+
+	if ms.updateWorkflowTaskVersionedTransition && ms.HasPendingWorkflowTask() {
+		ms.executionInfo.WorkflowTaskLastUpdatedVersionedTransition = currentVersionedTransition
```

Comment:
An alternative implementation here could be: set LastUpdatedVersionedTransition to nil when workflowTask is updated and when closing transaction, set it to the right value. 
However, that creates an issue when rolling out the new logic. Any sub state machine created before the new code is deployed will have the field set to nil and be treated the same way as those got updated in the current transaction.